src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/MultiTypePropertyFetcher.cs

```diff
@@ -0,0 +1,117 @@
+﻿// <copyright file="MultiTypePropertyFetcher.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Reflection;
+
+namespace OpenTelemetry.Instrumentation
+{
+    /// <summary>
+    /// PropertyFetcher fetches a property from an object.
+    /// </summary>
+    /// <typeparam name="T">The type of the property being fetched.</typeparam>
+    public class MultiTypePropertyFetcher<T>
+    {
+        private readonly string propertyName;
+        private ConcurrentDictionary<Type, PropertyFetch> innerFetcher = new ConcurrentDictionary<Type, PropertyFetch>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MultiTypePropertyFetcher{T}"/> class.
+        /// </summary>
+        /// <param name="propertyName">Property name to fetch.</param>
+        public MultiTypePropertyFetcher(string propertyName)
+        {
+            this.propertyName = propertyName;
+        }
+
+        /// <summary>
+        /// Fetch the property from the object.
+        /// </summary>
+        /// <param name="obj">Object to be fetched.</param>
+        /// <returns>Property fetched.</returns>
+        public T Fetch(object obj)
+        {
+            if (obj == null)
+            {
+                return default(T);
+            }
+
+            var type = obj.GetType().GetTypeInfo();
+            var fetcher = this.innerFetcher.GetOrAdd(type, t =>
```

Comment:
One small perf improvement you could make, don't use GetOrAdd. The lambda will always allocate a delegate (even if it isn't called). If you use TryGetValue and then TryAdd (if nothing was found) you can avoid that.

Reply:
Done